tests/handler/handler_copy.cpp

```diff
@@ -17,334 +22,887 @@ using mode_t = cl::sycl::access::mode;
+  template <typename test_fn>
+  void verify_update_host(test_fn fn, const log_helper& lh) const {
+    run_test_function(fn, lh);
+
+    std::lock_guard<cl::sycl::mutex_class> lock(srcBufHostMemoryMutex);
+    for (size_t i = 0; i < numElems; ++i) {
+      const auto idx = reconstruct_index(srcBufRange, i);
+      if (is_within_window(srcCopyOffset, srcCopyRange, idx)) {
+        const auto expected = encode_index_init_op<dataT, dim_src>{}(idx);
+        const auto received = srcBufHostMemory.get()[i];
+        if (!th::equal(received, expected)) {
+          log_error(lh, cl::sycl::id<3>(i, 0, 0), received, expected);
+          return;
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Verifies a host to device copy.
+   *
+   * While the device target may be strided, the source host memory region is
+   * always dense.
+   */
+  template <typename test_fn>
+  void verify_h2d_copy(test_fn fn, const log_helper& lh) {
+    run_test_function(fn, lh);
+
+    // TODO: Consider verifying directly on device.
+    auto acc = dstBuf->template get_access<cl::sycl::access::mode::read>();
+    for (size_t i = 0; i < numElems; ++i) {
+      const auto dstAbsIdx = reconstruct_index(dstBufRange, i);
+      const auto received = acc[dstAbsIdx];
+      if (is_within_window(dstCopyOffset, dstCopyRange, dstAbsIdx)) {
+        // Compute relative linear index within destination copy range.
+        const size_t relLinearIdx =
+            compute_relative_linear_id(dstCopyOffset, dstCopyRange, dstAbsIdx);
+
+        // SYCL doesn't support strided H2D copies, so this is dense.
+        const auto expected = srcHostPtr.get()[relLinearIdx];
+
+        if (!th::equal(received, expected)) {
+          log_error(lh, id_helper<3>::cast(dstAbsIdx), received, expected);
+          return;
+        }
+      } else {
+        if (!th::equal(received, deviceCanary)) {
+          log_canary_violation(lh, id_helper<3>::cast(dstAbsIdx), received);
+          return;
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Verifies a device to device copy.
+   */
+  template <typename test_fn>
+  void verify_d2d_copy(test_fn fn, const log_helper& lh) {
+    run_test_function(fn, lh);
+
+    // TODO: Consider verifying directly on device.
+    auto acc = dstBuf->template get_access<cl::sycl::access::mode::read>();
+    for (size_t i = 0; i < numElems; ++i) {
+      const auto dstAbsIdx = reconstruct_index(dstBufRange, i);
+      const auto received = acc[dstAbsIdx];
+      if (is_within_window(dstCopyOffset, dstCopyRange, dstAbsIdx)) {
+        // Compute relative linear index within destination copy range.
+        const size_t relLinearIdx =
+            compute_relative_linear_id(dstCopyOffset, dstCopyRange, dstAbsIdx);
+
+        // Now compute relative index in source copy range.
+        const auto srcRelIdx = reconstruct_index(srcCopyRange, relLinearIdx);
+
+        // Convert to absolute source index and use to compute expected value.
+        const auto expected =
```

Comment:
Factorizing the 2 functions and having a lambda expressing the 2 line differences?

Reply:
`verify_h2d_copy` and `verify_d2d_copy` you mean? I can look into it, sure!

Reply:
Yes they look pretty much the same.